extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/statistics/serde/KeyCollectorSnapshotSerializer.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.msq.statistics.serde;
+
+import org.apache.druid.msq.statistics.ClusterByStatisticsSnapshot;
+import org.apache.druid.msq.statistics.KeyCollectorSnapshot;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Serializes a {@link ClusterByStatisticsSnapshot} into a byte[].
+ */
+public abstract class KeyCollectorSnapshotSerializer
+{
+  protected abstract byte getType();
+
+  protected abstract byte[] serializeKeyCollector(KeyCollectorSnapshot collectorSnapshot);
+
+  public byte[] serialize(KeyCollectorSnapshot collectorSnapshot)
```

Comment:
You should also mention here that type cannot by the same across various implementation. 
Javadocs of the above serialized would be helpful.